### ISSUE TITLE
[BUG] When the asset name parameter is not a full URL, relative directory hash is evaluated from an empty string.

### ISSUE THREAD

**Issue by willianrampazzo** [avocado-framework/avocado#3659]
When an Asset object is created using the `name` parameter without a full URL, the method `_get_relative_dir` returns the hash of an empty string because `base_url = os.path.dirname(self.parsed_name.geturl())` evaluates to an empty string, as in the example below:

```
>>> os.path.dirname(urllib.parse.urlparse('foo.tgz').geturl())
''
>>> 
```

Previous code evaluated the hash of `:///` as it was "manually" building the URL.

This is not a consistent behavior, nor expected.

**Comment by willianrampazzo** [comment]
My suggestion to fix this problem is to move each piece of the URL to its correct place and do not mix asset name with locations.

With this suggestion, if an Asset object is created like:

```
self.hello = self.fetch_asset(
    name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
    locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
        'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
        'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
    )
```

- First, the file name should be removed from all the URLs in the `locations` parameter. 
- Second, the URL `https://mirrors.peers.community/mirrors/gnu/hello/` from `name` should be moved to the first index of `locations`.

It also means the relative directory built from the URL should match the first location where the `fetch` method was able to download the file.

**Comment by willianrampazzo** [comment]
Before I start the implementation, I think some discussion may be valid here. 

Just to add the context on how it works today, if the parameter `name` is **NOT** a URI and there is a `hash` parameter, the relative asset destination is `by_name`. Otherwise, the relative asset destination is `by_location` plus a subdirectory, created as the hash of URI in the `name` parameter without the asset name.

Let me describe some use cases for the discussion:

1. If the `name` parameter is a URI, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:

```
self.hello = self.fetch_asset(
    name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
    )
```

2. If the `name` parameter is a URI, but there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:

```
self.hello = self.fetch_asset(
    name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
    locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
        'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
        'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
    )
```

In this case, if the URI in the `name` parameter is unreachable and the asset is fetched from one of the other mirrors from `location`, the destination hash into `by_location` won't reflect the correct location. This problem impacts the `find_asset_file` method in the Asset class because it will also look for the asset in the directory representing the hash of the URI in the `name` parameter.

3. If the `name` parameter is **NOT** a URI, and there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of an empty string. The following code represents this:

```
self.hello = self.fetch_asset(
    name='hello-2.9.tar.gz',
    locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
        'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
        'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
    )
```

Again, we will have the same problem of an incorrect destination hash for the asset into `by_location` relative directory.

This third scenario may still be valid for a user fetching a package from a project that builds the same package name for its latest repository and send it to a couple of mirrors. In this case, every time the package is updated, the hash changes, so having the hash in the fetch command won't have the desired effect for the user.

The questions I have so far, and I think, deserve a discussion are:

1. Should we accept the use case described in **2**? If we agree it is a valid use case, I would like to use the suggestion from https://github.com/avocado-framework/avocado/issues/3659#issuecomment-599728305 to fix it.

2. Fixing the problem of having a hash for an empty string, use case **3** may lead to multiple destination subdirectories into `by_location` in case of the asset not being fetched from the first mirror in every attempt. Does it make sense to send files that have more than 1 location defined in the `locations` parameter to `by_name` relative directory, independent of having a hash for it?

Opinions and suggestions here are appreciated and welcome @clebergnu, @apahim!

**Comment by clebergnu** [comment]
> My suggestion to fix this problem is to move each piece of the URL to its correct place and do not mix asset name with locations.
> 
> With this suggestion, if an Asset object is created like:
> 
> ```
> self.hello = self.fetch_asset(
>     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
>     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
>         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
>         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
>     )
> ```
> 

Even though this scenario is currently supported (but IMO broken), I'd suggest we implement a clearer supported usage.  My suggestions:

* If `name` is *not* a URL, at least one entry in `locations` would be *required*
   * In this case, where name is a file name that was clearly given, the intended behavior is to get the file from any of the available location, so the location in the cache (currently implemented by `_get_relative_dir()` should *not* include the location. The goal is that the local cache location would not be different, and the file would be found in the cache, if the first location used once, and then another location is used later 
  * Note: there may be a caveat here, with odd cases when the last component of a location will not match the `name`, for instance, `https://github.com/avocado-framework/avocado/archive/77.0.tar.gz`.  In this case, should we add name to the location? We have to decide whether to support this or not

> * First, the file name should be removed from all the URLs in the `locations` parameter.

I generally agree, but see my previous comment.

> * Second, the URL `https://mirrors.peers.community/mirrors/gnu/hello/` from `name` should be moved to the first index of `locations`.
> 

Yes, I also agree this is simpler, but there's the caveat I noticed before.

> It also means the relative directory built from the URL should match the first location where the `fetch` method was able to download the file.

This point I don't agree.  The relative directory is just a tool to make sure we store and find the files in cache correctly.  So, if the file is the same, but available on multiple locations, the relative directory should be the same (because the file is the same) even if it's acquired from different locations.

**Comment by clebergnu** [comment]
> Before I start the implementation, I think some discussion may be valid here.
> 
> Just to add the context on how it works today, if the parameter `name` is **NOT** a URI and there is a `hash` parameter, the relative asset destination is `by_name`. Otherwise, the relative asset destination is `by_location` plus a subdirectory, created as the hash of URI in the `name` parameter without the asset name.
> 
> Let me describe some use cases for the discussion:
> 
> 1. If the `name` parameter is a URI, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> 
> ```
> self.hello = self.fetch_asset(
>     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
>     )
> ```
> 

So this use case is for once you have or know of a single possible location for a file. Given that the relative directory's goal is simply to store and find the file in the cache appropriately, I wonder why the location hash does *not* include the "file name" in the URI. I don't see why it shouldn't... can you think of a reason?

> 1. If the `name` parameter is a URI, but there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> 
> ```
> self.hello = self.fetch_asset(
>     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
>     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
>         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
>         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
>     )
> ```
> 
> In this case, if the URI in the `name` parameter is unreachable and the asset is fetched from one of the other mirrors from `location`, the destination hash into `by_location` won't reflect the correct location. This problem impacts the `find_asset_file` method in the Asset class because it will also look for the asset in the directory representing the hash of the URI in the `name` parameter.

I see, you're right. I can only think of changing the allowed uses of the `fetch_asset` asset method so that if multiple locations are given, a name should be just that, a file name.  Do you have other ideas?

> 
> 1. If the `name` parameter is **NOT** a URI, and there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of an empty string. The following code represents this:
> 

This sounds wrong to me, I mean, instead of `by_location` + `empty_hash`, it should just be `by_name`, which amounts to the same level of separation.

> ```
> self.hello = self.fetch_asset(
>     name='hello-2.9.tar.gz',
>     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
>         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
>         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
>     )
> ```
> 
> Again, we will have the same problem of an incorrect destination hash for the asset into `by_location` relative directory.
> 

Yeah, I don't think that's intended.

> This third scenario may still be valid for a user fetching a package from a project that builds the same package name for its latest repository and send it to a couple of mirrors. In this case, every time the package is updated, the hash changes, so having the hash in the fetch command won't have the desired effect for the user.
> 
> The questions I have so far, and I think, deserve a discussion are:
> 
> 1. Should we accept the use case described in **2**? If we agree it is a valid use case, I would like to use the suggestion from [#3659 (comment)](https://github.com/avocado-framework/avocado/issues/3659#issuecomment-599728305) to fix it.

I think that is broken.  Like I said elsewhere, I think that use case should be one that accepts `name` is a file name *only* and requires one or more locations.

> 2. Fixing the problem of having a hash for an empty string, use case **3** may lead to multiple destination subdirectories into `by_location` in case of the asset not being fetched from the first mirror in every attempt. Does it make sense to send files that have more than 1 location defined in the `locations` parameter to `by_name` relative directory, independent of having a hash for it?
> 

That's what I would expect!

> Opinions and suggestions here are appreciated and welcome @clebergnu, @apahim!

**Comment by willianrampazzo** [comment]
> > Before I start the implementation, I think some discussion may be valid here.
> > Just to add the context on how it works today, if the parameter `name` is **NOT** a URI and there is a `hash` parameter, the relative asset destination is `by_name`. Otherwise, the relative asset destination is `by_location` plus a subdirectory, created as the hash of URI in the `name` parameter without the asset name.
> > Let me describe some use cases for the discussion:
> > 
> > 1. If the `name` parameter is a URI, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> > 
> > ```
> > self.hello = self.fetch_asset(
> >     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
> >     )
> > ```
> 
> So this use case is for once you have or know of a single possible location for a file. Given that the relative directory's goal is simply to store and find the file in the cache appropriately, I wonder why the location hash does _not_ include the "file name" in the URI. I don't see why it shouldn't... can you think of a reason?

There would be just one reason I can think of if I am a user using this interface. In case I was fetching multiple files from the same URI, I would expect all the fetched files to be located in the same cache directory. It means that all files fetched from, for example, `https://mirrors.kernel.org/gnu/hello/`, reside on the same cache location, but files fetched from `https://mirrors.kernel.org/gnu/git/` reside on its cache location. In my understanding, the local hash for the location should be the full URI, except the file name. Does that make sense?

> 
> > 1. If the `name` parameter is a URI, but there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> > 
> > ```
> > self.hello = self.fetch_asset(
> >     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
> >     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
> >         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
> >         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
> >     )
> > ```
> > 
> > 
> > In this case, if the URI in the `name` parameter is unreachable and the asset is fetched from one of the other mirrors from `location`, the destination hash into `by_location` won't reflect the correct location. This problem impacts the `find_asset_file` method in the Asset class because it will also look for the asset in the directory representing the hash of the URI in the `name` parameter.
> 
> I see, you're right. I can only think of changing the allowed uses of the `fetch_asset` asset method so that if multiple locations are given, a name should be just that, a file name. Do you have other ideas?

It should not be difficult to move the URI in the `name` parameter to the `locations` list instead of forcing the location to the in the correct parameter. It keeps the API flexible. Or we can just restrict the use cases of `name` and `locations` as I listed at the end of this comment.

> 
> > 1. If the `name` parameter is **NOT** a URI, and there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of an empty string. The following code represents this:
> 
> This sounds wrong to me, I mean, instead of `by_location` + `empty_hash`, it should just be `by_name`, which amounts to the same level of separation.

Agree!

> 
> > ```
> > self.hello = self.fetch_asset(
> >     name='hello-2.9.tar.gz',
> >     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
> >         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
> >         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
> >     )
> > ```
> > 
> > 
> > Again, we will have the same problem of an incorrect destination hash for the asset into `by_location` relative directory.
> 
> Yeah, I don't think that's intended.
> 
> > This third scenario may still be valid for a user fetching a package from a project that builds the same package name for its latest repository and send it to a couple of mirrors. In this case, every time the package is updated, the hash changes, so having the hash in the fetch command won't have the desired effect for the user.
> > The questions I have so far, and I think, deserve a discussion are:
> > 
> > 1. Should we accept the use case described in **2**? If we agree it is a valid use case, I would like to use the suggestion from [#3659 (comment)](https://github.com/avocado-framework/avocado/issues/3659#issuecomment-599728305) to fix it.
> 
> I think that is broken. Like I said elsewhere, I think that use case should be one that accepts `name` is a file name _only_ and requires one or more locations.

So, in a nutshell we have 2 possibilities:

1. `name` should be a full URI and `locations` should be empty.
2. `name` should be a single file name and `locations` should have one or more entries.

> 
> > 1. Fixing the problem of having a hash for an empty string, use case **3** may lead to multiple destination subdirectories into `by_location` in case of the asset not being fetched from the first mirror in every attempt. Does it make sense to send files that have more than 1 location defined in the `locations` parameter to `by_name` relative directory, independent of having a hash for it?
> 
> That's what I would expect!

Agree!

> 
> > Opinions and suggestions here are appreciated and welcome @clebergnu, @apahim!

**Comment by clebergnu** [comment]
> > > Before I start the implementation, I think some discussion may be valid here.
> > > Just to add the context on how it works today, if the parameter `name` is **NOT** a URI and there is a `hash` parameter, the relative asset destination is `by_name`. Otherwise, the relative asset destination is `by_location` plus a subdirectory, created as the hash of URI in the `name` parameter without the asset name.
> > > Let me describe some use cases for the discussion:
> > > 
> > > 1. If the `name` parameter is a URI, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> > > 
> > > ```
> > > self.hello = self.fetch_asset(
> > >     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
> > >     )
> > > ```
> > 
> > 
> > So this use case is for once you have or know of a single possible location for a file. Given that the relative directory's goal is simply to store and find the file in the cache appropriately, I wonder why the location hash does _not_ include the "file name" in the URI. I don't see why it shouldn't... can you think of a reason?
> 
> There would be just one reason I can think of if I am a user using this interface. In case I was fetching multiple files from the same URI, I would expect all the fetched files to be located in the same cache directory. It means that all files fetched from, for example, `https://mirrors.kernel.org/gnu/hello/`, reside on the same cache location, but files fetched from `https://mirrors.kernel.org/gnu/git/` reside on its cache location. In my understanding, the local hash for the location should be the full URI, except the file name. Does that make sense?
> 

About all files coming from the same URL being stored in the same directory, I have the feeling that this is much more of an implementation detail. It seems nice to have that indeed, allowing users eventually to navigate that directory and its contents, but it seems like a secondary goal. With regards to removing the file name from the hash and using only the base location (such as `https://mirrors.kernel.org/gnu/hello/`) I can not think of any problem, given that we also expect that the remaining URL components (the "file name") will also be unique.  So, to summarize the answer, I think it does make sense. Hopefully we won't be disproved by the end result! :) 

> > > 1. If the `name` parameter is a URI, but there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of the URI in the `name` parameter without the file name. The following code represents this:
> > > 
> > > ```
> > > self.hello = self.fetch_asset(
> > >     name='https://mirrors.peers.community/mirrors/gnu/hello/hello-2.9.tar.gz',
> > >     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
> > >         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
> > >         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
> > >     )
> > > ```
> > > 
> > > 
> > > In this case, if the URI in the `name` parameter is unreachable and the asset is fetched from one of the other mirrors from `location`, the destination hash into `by_location` won't reflect the correct location. This problem impacts the `find_asset_file` method in the Asset class because it will also look for the asset in the directory representing the hash of the URI in the `name` parameter.
> > 
> > 
> > I see, you're right. I can only think of changing the allowed uses of the `fetch_asset` asset method so that if multiple locations are given, a name should be just that, a file name. Do you have other ideas?
> 
> It should not be difficult to move the URI in the `name` parameter to the `locations` list instead of forcing the location to the in the correct parameter. It keeps the API flexible. Or we can just restrict the use cases of `name` and `locations` as I listed at the end of this comment.
> 
> > > 1. If the `name` parameter is **NOT** a URI, and there are mirrors listed in the `location` parameter, the relative directory will be `by_location`, and the subdirectory into it will be the hash of an empty string. The following code represents this:
> > 
> > 
> > This sounds wrong to me, I mean, instead of `by_location` + `empty_hash`, it should just be `by_name`, which amounts to the same level of separation.
> 
> Agree!
> 
> > > ```
> > > self.hello = self.fetch_asset(
> > >     name='hello-2.9.tar.gz',
> > >     locations=['https://mirrors.kernel.org/gnu/hello/hello-2.9.tar.gz',
> > >         'http://gnu.c3sl.ufpr.br/ftp/hello-2.9.tar.gz',
> > >         'ftp://ftp.funet.fi/pub/gnu/prep/hello/hello-2.9.tar.gz']
> > >     )
> > > ```
> > > 
> > > 
> > > Again, we will have the same problem of an incorrect destination hash for the asset into `by_location` relative directory.
> > 
> > 
> > Yeah, I don't think that's intended.
> > > This third scenario may still be valid for a user fetching a package from a project that builds the same package name for its latest repository and send it to a couple of mirrors. In this case, every time the package is updated, the hash changes, so having the hash in the fetch command won't have the desired effect for the user.
> > > The questions I have so far, and I think, deserve a discussion are:
> > > 
> > > 1. Should we accept the use case described in **2**? If we agree it is a valid use case, I would like to use the suggestion from [#3659 (comment)](https://github.com/avocado-framework/avocado/issues/3659#issuecomment-599728305) to fix it.
> > 
> > 
> > I think that is broken. Like I said elsewhere, I think that use case should be one that accepts `name` is a file name _only_ and requires one or more locations.
> 
> So, in a nutshell we have 2 possibilities:
> 
> 1. `name` should be a full URI and `locations` should be empty.
> 2. `name` should be a single file name and `locations` should have one or more entries.
> 

I think that's fair enough, and removes a good deal of confusion on the expected behavior.

> > > 1. Fixing the problem of having a hash for an empty string, use case **3** may lead to multiple destination subdirectories into `by_location` in case of the asset not being fetched from the first mirror in every attempt. Does it make sense to send files that have more than 1 location defined in the `locations` parameter to `by_name` relative directory, independent of having a hash for it?
> > 
> > 
> > That's what I would expect!
> 
> Agree!
> 
> > > Opinions and suggestions here are appreciated and welcome @clebergnu, @apahim!
